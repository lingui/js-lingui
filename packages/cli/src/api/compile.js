// @flow
import * as t from "babel-types"
import { parse } from "messageformat-parser"
import { parseExpression } from "babylon"
import generate from "babel-generator"
import plurals from "make-plural"
import R from "ramda"

import type { CatalogType } from "./formats/types"

const isString = s => typeof s === "string"

export function compile(message: string) {
  function processTokens(tokens) {
    if (!tokens.filter(token => !isString(token)).length) {
      return tokens.join("")
    }

    return t.arrayExpression(
      tokens.map(token => {
        if (isString(token)) {
          return t.stringLiteral(token)

          // # in plural case
        } else if (token.type === "octothorpe") {
          return t.stringLiteral("#")

          // simple argument
        } else if (token.type === "argument") {
          return t.callExpression(arg, [t.stringLiteral(token.arg)])

          // argument with custom format (date, number)
        } else if (token.type === "function") {
          const params = [
            t.stringLiteral(token.arg),
            t.stringLiteral(token.key)
          ]

          const format = token.params[0]
          if (format) {
            params.push(t.stringLiteral(format))
          }
          return t.callExpression(arg, params)
        }

        // complex argument with cases
        const formatProps = []

        if (token.offset) {
          formatProps.push(
            t.objectProperty(
              t.identifier("offset"),
              t.numericLiteral(parseInt(token.offset))
            )
          )
        }

        token.cases.forEach(item => {
          const inlineTokens = processTokens(item.tokens)
          formatProps.push(
            t.objectProperty(
              t.identifier(item.key),
              isString(inlineTokens)
                ? t.stringLiteral(inlineTokens)
                : inlineTokens
            )
          )
        })

        const params = [
          t.stringLiteral(token.arg),
          t.stringLiteral(token.type),
          t.objectExpression(formatProps)
        ]

        return t.callExpression(arg, params)
      })
    )
  }

  const arg = t.identifier("a")

  const tokens = parse(message)
  const ast = processTokens(tokens)

  if (isString(ast)) return t.stringLiteral(ast)

  return t.functionExpression(
    null,
    [arg],
    t.blockStatement([t.returnStatement(ast)])
  )
}

export function createCompiledCatalog(locale: string, messages: CatalogType) {
  const [language] = locale.split(/[_-]/)
  const pluralRules = plurals[language]

  const compiledMessages = R.keys(messages).map(key => {
    const translation = messages[key]
    return t.objectProperty(t.stringLiteral(key), compile(translation || key))
  })

  const languageData = [
    t.objectProperty(
      t.stringLiteral("plurals"),
      parseExpression(pluralRules.toString())
    )
  ]

  const ast = t.expressionStatement(
    t.assignmentExpression(
      "=",
      t.memberExpression(t.identifier("module"), t.identifier("exports")),
      t.objectExpression([
        // language data
        t.objectProperty(
          t.identifier("languageData"),
          t.objectExpression(languageData)
        ),
        // messages
        t.objectProperty(
          t.identifier("messages"),
          t.objectExpression(compiledMessages)
        )
      ])
    )
  )

  return generate(ast, {
    minified: true
  }).code
}
