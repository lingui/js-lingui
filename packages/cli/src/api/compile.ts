import * as t from "@babel/types"
import generate, { GeneratorOptions } from "@babel/generator"
import {
  compileMessage,
  CompiledMessage,
} from "@lingui/message-utils/compileMessage"
import pseudoLocalize from "./pseudoLocalize"

export type CompiledCatalogNamespace = "cjs" | "es" | "ts" | "json" | string

type CompiledCatalogType = {
  [msgId: string]: string
}

export type CreateCompileCatalogOptions = {
  strict?: boolean
  namespace?: CompiledCatalogNamespace
  pseudoLocale?: string
  compilerBabelOptions?: GeneratorOptions
}

export function createCompiledCatalog(
  locale: string,
  messages: CompiledCatalogType,
  options: CreateCompileCatalogOptions
): string {
  const {
    strict = false,
    namespace = "cjs",
    pseudoLocale,
    compilerBabelOptions = {},
  } = options
  const shouldPseudolocalize = locale === pseudoLocale

  const compiledMessages = Object.keys(messages).reduce<{
    [msgId: string]: CompiledMessage
  }>((obj, key: string) => {
    // Don't use `key` as a fallback translation in strict mode.
    const translation = (messages[key] || (!strict ? key : "")) as string

    obj[key] = compile(translation, shouldPseudolocalize)
    return obj
  }, {})

  if (namespace === "json") {
    return JSON.stringify({ messages: compiledMessages })
  }

  const ast = buildExportStatement(
    //build JSON.parse(<compiledMessages>) statement
    t.callExpression(
      t.memberExpression(t.identifier("JSON"), t.identifier("parse")),
      [t.stringLiteral(JSON.stringify(compiledMessages))]
    ),
    namespace
  )

  const code = generate(ast, {
    minified: true,
    jsescOption: {
      minimal: true,
    },
    ...compilerBabelOptions,
  }).code

  return "/*eslint-disable*/" + code
}

function buildExportStatement(
  expression: t.Expression,
  namespace: CompiledCatalogNamespace
) {
  if (namespace === "es" || namespace === "ts") {
    // export const messages = { message: "Translation" }
    return t.exportNamedDeclaration(
      t.variableDeclaration("const", [
        t.variableDeclarator(t.identifier("messages"), expression),
      ])
    )
  } else {
    let exportExpression = null
    const matches = namespace.match(/^(window|global)\.([^.\s]+)$/)
    if (namespace === "cjs") {
      // module.exports.messages = { message: "Translation" }
      exportExpression = t.memberExpression(
        t.identifier("module"),
        t.identifier("exports")
      )
    } else if (matches) {
      // window.i18nMessages = { messages: { message: "Translation" }}
      exportExpression = t.memberExpression(
        t.identifier(matches[1]),
        t.identifier(matches[2])
      )
    } else {
      throw new Error(`Invalid namespace param: "${namespace}"`)
    }

    return t.expressionStatement(
      t.assignmentExpression(
        "=",
        exportExpression,
        t.objectExpression([
          t.objectProperty(t.identifier("messages"), expression),
        ])
      )
    )
  }
}

/**
 * Compile string message into AST tree. Message format is parsed/compiled into
 * JS arrays, which are handled in client.
 */
export function compile(
  message: string,
  shouldPseudolocalize: boolean = false
) {
  return compileMessage(message, (value) =>
    shouldPseudolocalize ? pseudoLocalize(value) : value
  )
}
