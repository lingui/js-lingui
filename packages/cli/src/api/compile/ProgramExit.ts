export class ProgramExit extends Error {}
