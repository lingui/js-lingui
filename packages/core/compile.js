/**
 * This entry is for old runtimes which do not support `exports` field in package.json
 * https://github.com/facebook/metro/issues/670
 */
module.exports = require("./build/cjs/compile.js")
