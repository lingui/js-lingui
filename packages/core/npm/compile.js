module.exports = require("./cjs/compile.js")
