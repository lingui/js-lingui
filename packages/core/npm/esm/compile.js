import {
    compileMessage as compileMessageProd
} from './compile.production.min.js';

import {
    compileMessage as compileMessageDev
} from './compile.development.js';

export const compileMessage = process.env.NODE_ENV === 'production' ? compileMessageProd : compileMessageDev;
