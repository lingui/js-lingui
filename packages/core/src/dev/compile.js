// @flow
import { parse } from "messageformat-parser"
import { isString } from "../essentials"

// [Tokens] -> (CTX -> String)
function processTokens(tokens, octothorpe = {}): Function => string {
  if (!tokens.filter(token => !isString(token)).length) {
    return tokens.join("")
  }

  return ctx =>
    tokens.map(token => {
      if (isString(token)) {
        return token

        // # in plural case
      } else if (token.type === "octothorpe") {
        const { name, offset = 0 } = octothorpe
        return ctx(name) - offset

        // simple argument
      } else if (token.type === "argument") {
        return ctx(token.arg)

        // argument with custom format (date, number)
      } else if (token.type === "function") {
        return ctx(token.arg, token.key, token.params[0])
      }

      const offset = token.offset ? parseInt(token.offset) : undefined

      // complex argument with cases
      const formatProps = {}
      token.cases.forEach(item => {
        formatProps[item.key] = processTokens(item.tokens, {
          name: token.arg,
          offset
        })
      })

      return ctx(token.arg, token.type, {
        offset,
        ...formatProps
      })
    })
}

// Message -> (Params -> String)
export default function compile(message: string): Function | string {
  try {
    return processTokens(parse(message))
  } catch (e) {
    console.error(`Message cannot be parsed due to syntax erorrs: ${message}`)
    return message
  }
}
