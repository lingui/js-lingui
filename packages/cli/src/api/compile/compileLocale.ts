import { writeCompiled } from "../catalog"
import { LinguiConfigNormalized } from "@lingui/conf"
import pico from "picocolors"
import { getCatalogs, getMergedCatalogPath } from "../catalog/getCatalogs"
import { CliCompileOptions } from "../../lingui-compile"
import { ProgramExit } from "../ProgramExit"
import { createCompiledCatalog } from "../compile"

import normalizePath from "normalize-path"
import nodepath from "path"
import { createCompilationErrorMessage } from "../messages"
import { getTranslationsForCatalog } from "../catalog/getTranslationsForCatalog"
import { Logger } from "../logger"

export async function compileLocale(
  locale: string,
  options: CliCompileOptions,
  config: LinguiConfigNormalized,
  doMerge: boolean,
  logger: Logger
) {
  const catalogs = await getCatalogs(config)

  let mergedCatalogs: Record<string, string> = {}

  for (const catalog of catalogs) {
    const { messages, missing: missingMessages } =
      await getTranslationsForCatalog(catalog, locale, {
        fallbackLocales: config.fallbackLocales,
        sourceLocale: config.sourceLocale,
      })

    if (
      !options.allowEmpty &&
      locale !== config.pseudoLocale &&
      missingMessages.length > 0
    ) {
      logger.error(
        pico.red(
          `Error: Failed to compile catalog for locale ${pico.bold(locale)}!`
        )
      )

      if (options.verbose) {
        logger.error(pico.red("Missing translations:"))
        missingMessages.forEach((missing) => {
          const source =
            missing.source || missing.source === missing.id
              ? `: (${missing.source})`
              : ""

          logger.error(`${missing.id}${source}`)
        })
      } else {
        logger.error(
          pico.red(`Missing ${missingMessages.length} translation(s)`)
        )
      }
      logger.error()
      throw new ProgramExit()
    }

    if (doMerge) {
      mergedCatalogs = { ...mergedCatalogs, ...messages }
    } else {
      if (
        !(await compileAndWrite(
          locale,
          config,
          options,
          catalog.path,
          messages,
          logger
        ))
      ) {
        throw new ProgramExit()
      }
    }
  }

  if (doMerge) {
    const result = await compileAndWrite(
      locale,
      config,
      options,
      await getMergedCatalogPath(config),
      mergedCatalogs,
      logger
    )

    if (!result) {
      throw new ProgramExit()
    }
  }
}

async function compileAndWrite(
  locale: string,
  config: LinguiConfigNormalized,
  options: CliCompileOptions,
  writePath: string,
  messages: Record<string, string>,
  logger: Logger
): Promise<boolean> {
  const namespace = options.typescript
    ? "ts"
    : options.namespace || config.compileNamespace
  const { source: compiledCatalog, errors } = createCompiledCatalog(
    locale,
    messages,
    {
      strict: false,
      namespace,
      pseudoLocale: config.pseudoLocale,
      compilerBabelOptions: config.compilerBabelOptions,
    }
  )

  if (errors.length) {
    let message = createCompilationErrorMessage(locale, errors)

    if (options.failOnCompileError) {
      message += `These errors fail command execution because \`--strict\` option passed`
      logger.error(pico.red(message))
      return false
    } else {
      message += `You can fail command execution on these errors by passing \`--strict\` option`
      logger.error(pico.red(message))
    }
  }

  let compiledPath = await writeCompiled(
    writePath,
    locale,
    compiledCatalog,
    namespace
  )

  compiledPath = normalizePath(nodepath.relative(config.rootDir, compiledPath))

  options.verbose && logger.error(pico.green(`${locale} â‡’ ${compiledPath}`))
  return true
}
