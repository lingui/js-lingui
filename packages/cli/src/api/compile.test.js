import generate from "babel-generator"
import { compile, createCompiledCatalog } from "./compile"

describe("compile", function() {
  const getSource = message =>
    generate(compile(message), {
      compact: true,
      minified: true
    }).code

  it("should optimize string only messages", function() {
    expect(getSource("Hello World")).toEqual('"Hello World"')
  })

  it("should compile arguments", function() {
    expect(getSource("{name}")).toEqual('function(a){return[a("name")]}')

    expect(getSource("B4 {name} A4")).toEqual(
      'function(a){return["B4 ",a("name")," A4"]}'
    )
  })

  it("should compile arguments with formats", function() {
    expect(getSource("{name, number}")).toEqual(
      'function(a){return[a("name","number")]}'
    )

    expect(getSource("{name, number, percent}")).toEqual(
      'function(a){return[a("name","number","percent")]}'
    )
  })

  it("should compile plural", function() {
    expect(getSource("{name, plural, one {Book} other {Books}}")).toEqual(
      'function(a){return[a("name","plural",{one:"Book",other:"Books"})]}'
    )

    expect(
      getSource("{name, plural, one {Book} other {{name} Books}}")
    ).toEqual(
      'function(a){return[a("name","plural",{one:"Book",other:[a("name")," Books"]})]}'
    )

    expect(getSource("{name, plural, one {Book} other {# Books}}")).toEqual(
      'function(a){return[a("name","plural",{one:"Book",other:["#"," Books"]})]}'
    )

    expect(
      getSource(
        "{name, plural, offset:1 =0 {No Books} one {Book} other {# Books}}"
      )
    ).toEqual(
      'function(a){return[a("name","plural",{offset:1,0:"No Books",one:"Book",other:["#"," Books"]})]}'
    )
  })

  it("should compile select", function() {
    expect(getSource("{name, select, male {He} female {She}}")).toEqual(
      'function(a){return[a("name","select",{male:"He",female:"She"})]}'
    )

    expect(getSource("{name, select, male {He} female {{name} She}}")).toEqual(
      'function(a){return[a("name","select",{male:"He",female:[a("name")," She"]})]}'
    )

    expect(getSource("{name, select, male {He} female {# She}}")).toEqual(
      'function(a){return[a("name","select",{male:"He",female:"# She"})]}'
    )
  })

  it("should report failed message on error", function() {
    expect(() =>
      getSource("{value, plural, one {Book} other {Books")
    ).toThrowErrorMatchingSnapshot()
  })
})

describe("createCompiledCatalog", function() {
  describe("namespace", function() {
    const getCompiledCatalog = (...args) =>
      createCompiledCatalog("fr", {}, ...args)
    it("should compile with cjs by default", function() {
      expect(getCompiledCatalog()).toMatchSnapshot()
    })

    it("should compile with es", function() {
      expect(getCompiledCatalog(false, "es")).toMatchSnapshot()
    })

    it("should compile with window", function() {
      expect(getCompiledCatalog(false, "window.test")).toMatchSnapshot()
    })

    it("should compile with global", function() {
      expect(getCompiledCatalog(false, "global.test")).toMatchSnapshot()
    })

    it("should error with invalid value", function() {
      expect(() =>
        getCompiledCatalog(false, "global")
      ).toThrowErrorMatchingSnapshot()
    })
  })
})
