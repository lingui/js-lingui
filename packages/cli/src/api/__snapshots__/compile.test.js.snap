// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compile should report failed message on error 1`] = `Can't parse message. Please check correct syntax: "{value, plural, one {Book} other {Books"`;

exports[`createCompiledCatalog namespace should compile with cjs by default 1`] = `/* eslint-disable */module.exports={languageData:{"plurals":function(n,ord){if(ord)return n==1?"one":"other";return n>=0&&n<2?"one":"other"}},messages:{}};`;

exports[`createCompiledCatalog namespace should compile with es 1`] = `/* eslint-disable */export default{languageData:{"plurals":function(n,ord){if(ord)return n==1?"one":"other";return n>=0&&n<2?"one":"other"}},messages:{}};`;

exports[`createCompiledCatalog namespace should compile with global 1`] = `/* eslint-disable */global.test={languageData:{"plurals":function(n,ord){if(ord)return n==1?"one":"other";return n>=0&&n<2?"one":"other"}},messages:{}};`;

exports[`createCompiledCatalog namespace should compile with window 1`] = `/* eslint-disable */window.test={languageData:{"plurals":function(n,ord){if(ord)return n==1?"one":"other";return n>=0&&n<2?"one":"other"}},messages:{}};`;

exports[`createCompiledCatalog namespace should error with invalid value 1`] = `Invalid namespace param: "global"`;
